spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java

```diff
@@ -327,19 +340,18 @@ public Boolean execute(RedisOperations ops) throws DataAccessException {
 
 					ops.watch(key); //monitor key
 
-					ops.multi(); //transaction start
+					if (ops.opsForValue().get(key) != null) {
+						ops.unwatch(); //key already exists, stop monitoring
+						return false;
+					}
 
-					//can't rely on operations result inside transaction, execution is delayed till `exec()`
-					ops.opsForValue().setIfAbsent(key, RedisLock.this);
+					ops.multi(); //transaction start
 
-					//set expire on key if exists
-					ops.expire(key, RedisLockRegistry.this.expireAfter, TimeUnit.MILLISECONDS);
+					//set the value and expire
+					ops.opsForValue().set(key, RedisLock.this, RedisLockRegistry.this.expireAfter, TimeUnit.MILLISECONDS);
 
 					//exec will contain all operations result or null - if execution has been aborted due to 'watch'
-					List result = ops.exec();
-
-					//check 'setIfAbsent' result (first in list)
-					return (result != null) && (!result.isEmpty()) && (Boolean.TRUE.equals(result.get(0)));
+					return ops.exec() != null;
```

Comment:
Hello, @kiakimov !

Glad to see your contribution one more time. 
And that's great that you fix not only our bugs but yours, too :smile: 

Well, I don't mind for that `reLock` fix, but I need some explanation to this change, please.

Previously we relied on the atomic `setIfAbsent` to be sure that only the single process populates the value.
Right, we have modified `expire` previously unconditionally, but right now you do that for whole `key`.
Does `WATCH` guaranty that only our process does changes?
Should I believe to this doc: http://redis.io/topics/transactions ?

I addition: how about test-case to cover the issue?

From other side would be better if you'd provide comprehensive commit message: https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.md. See the last paragraph.

Thank you very much anyway!